Add crossgen2 automated SuperPMI collection of Core_Root libraries

BruceForstall opened the pull request:
Add a superpmi.py collect `-exclude` option to avoid collecting
assemblies known to fail (and cause pop-up failures: see #47552).

Use `-exclude` to exclude R2RTest\Microsoft.Build.dll.

Depends on #47779

Fixes #47541

BruceForstall committed:
Add crossgen2 automated SuperPMI collection of Core_Root libraries

Exclude R2RTest\Microsoft.Build.dll from automated collection.

Add a superpmi.py collect `-exclude` option to avoid collecting
assemblies known to fail (and cause pop-up failures: see #47552).
(This isn't used by the automated collection, but is generally useful.)

Fixes #47541

BruceForstall commented:
@dotnet/jit-contrib 
cc @nattress

BruceForstall commented:
cc @dotnet/crossgen-contrib

kunalspathak commented on src/coreclr/scripts/superpmi.py:
```diff
@@ -244,7 +244,8 @@
 collect_parser.add_argument("--crossgen2", action="store_true", help="Run crossgen2 on a set of directories or assemblies.")
-collect_parser.add_argument("-assemblies", dest="assemblies", nargs="+", default=[], help="Pass a sequence of managed dlls or directories to recursively use while collecting with PMI, crossgen, or crossgen2. Required if --pmi, --crossgen, or --crossgen2 is specified.")
+collect_parser.add_argument("-assemblies", dest="assemblies", nargs="+", default=[], help="A list of managed dlls or directories to recursively use while collecting with PMI, crossgen, or crossgen2. Required if --pmi, --crossgen, or --crossgen2 is specified.")
+collect_parser.add_argument("-exclude", dest="exclude", nargs="+", default=[], help="A list of files or directories to exclude from the files and directories specified by `-assemblies`.")
```
I would recommend the right way to do this would be to not copy the assemblies that are problematic to the `payload` directory. `superpmi-setup.py` already has an ability to do that:
https://github.com/dotnet/runtime/blob/4de28f317c227b433fafecd5496bae04c2bca007/src/coreclr/scripts/superpmi-setup.py#L296

Perhaps, just add a condition there that if the `collection_name == 'crossgen2'`, exclude it.

kunalspathak replied:
With that, you won't need any changes in `superpmi.py`. The script will just act on the `assemblies` you provided.

BruceForstall replied:
ok, I moved the exclusion to superpmi-setup.py. The superpmi.py `-exclude` code is not required now, but I'd like to keep it as it could be generally useful.

kunalspathak reviewed (commented):
Overall looks good (as long as superpmi collection pipeline is green) except a comment about moving `--exclude` from `superpmi.py` to `superpmi-setup.py`.

kunalspathak reviewed (approved):
LGTM. Thanks!

kunalspathak commented:
I have kicked off a superpmi collection run based on this PR - https://dev.azure.com/dnceng/internal/_build/results?buildId=978167&view=results

kunalspathak commented:
> I have kicked off a superpmi collection run based on this PR - https://dev.azure.com/dnceng/internal/_build/results?buildId=978167&view=results

Never mind. I just saw that it depends on https://github.com/dotnet/runtime/pull/47779. I have cancelled the run.

BruceForstall commented:
I kicked off https://dev.azure.com/dnceng/internal/_build/results?buildId=978328&view=results

BruceForstall committed:
Copy .dotnet directory to correlation directory

Add .dotnet directory to path, so it can be found by superpmi.py when
setting up a crossgen2 collection.

Extra: if `-temp_dir` is specified to superpmi.py, normalize it to
an absolute path.

BruceForstall committed:
Only add .dotnet directory to path for crossgen2 collection

BruceForstall committed:
Try different HelixPreCommands syntax

BruceForstall committed:
Add `-collection_type` argument to superpmi-setup.py

BruceForstall committed:
Fix PATH variable separator for Linux in superpmi.py

Fix Windows PATH setting XML separator escape character

Add some more logging output

BruceForstall committed:
More spmi fixes

1. exclude more native files from superpmi setup
2. write stdout and stderr files to collection log on failure
3. write superpmi collection error return code to log on failure
4. fix creation of log file directories when `-log_file` argument specified
5. stop logging too-verbose message about CORINFO_SIG_INST handle
array illegal values, since it's "by design" for crossgen2

BruceForstall committed:
Pass `-log_level DEBUG` to all superpmi.py invocations in CI

This makes it easier to see the log output without needing to
find and download the logfile separately from the CI.

BruceForstall committed:
Generalize copy_directory() function

Allow copying .json files, which is required by crossgen2.

Don't filter anything in copy of .dotnet directory; I don't know
what kind of files are there, and don't want to filter any of them.

BruceForstall commented:
Latest: https://dev.azure.com/dnceng/internal/_build/results?buildId=981297&view=results

BruceForstall commented:
@kunalspathak I've made a bunch of additional changes here (you've probably noticed), in case you want to review them.

kunalspathak commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -356,9 +380,17 @@ def main(main_args):
+        # Need to accept files without any extension, which is how executable filesnames look.
+        acceptable_copy = lambda path: (os.path.basename(path).find(".") == -1) or any(path.endswith(extension) for extension in [".py", ".dll", ".so", ".json"])
+
+    print("is_windows {}".format(is_windows))
```
Delete.

kunalspathak commented on src/coreclr/scripts/superpmi-setup.py:
```diff
@@ -50,23 +50,40 @@
     "clretwrc.dll",
     "clrgc.dll",
     "clrjit.dll",
+    "clrjit_unix_arm_arm.dll",
```
Thanks for doing this. Wanted to do from long time.

kunalspathak reviewed (approved):
LGTM, with a minor suggestion.

BruceForstall committed:
Exclude more files

BruceForstall committed:
Use corerun instead of .dotnet so cross-plat works

BruceForstall committed:
Lower-case JIT-EE version string

BruceForstall committed:
Disable Linux crossgen2 jobs

They crash during collection and need to be investigated.
